### ISSUE TITLE
Same illegal memory access issue from #68

### ISSUE THREAD

**Issue by meder411** [colmap/colmap#79]
Hey Johannes,

I updated the colmap build with the most recent changes, and I still get the errors reported in https://github.com/colmap/colmap/issues/68.

I am using 126 4k images sampled from a video for my reconstruction and running exhaustive matching. Below is a snippet of the feature extraction steps from running the command:

```
feature_extractor --General.database_path COLMAP_DB_PATH --General.image_path FRAMES_DIR  --ExtractionOptions.reader_camera_model SIMPLE_RADIAL --ExtractionOptions.reader_single_camera 1 --ExtractionOptions.sift_max_num_features 40000 --ExtractionOptions.sift_max_image_size 4000
```

```
Processing file [121/126]
  Name:           img_002400.png
  Width:          3840px
  Height:         2160px
  Camera ID:      1
  Camera Model:   SIMPLE_RADIAL
  Features:       31175
Processing file [122/126]
  Name:           img_002420.png
  Width:          3840px
  Height:         2160px
  Camera ID:      1
  Camera Model:   SIMPLE_RADIAL
  Features:       32733
Processing file [123/126]
  Name:           img_002440.png
  Width:          3840px
  Height:         2160px
  Camera ID:      1
  Camera Model:   SIMPLE_RADIAL
  Features:       34514
```

As you can see we have some serious feature overload. I then run the exhaustive matcher

```
exhaustive_matcher --General.database_path COLMAP_DB_PATH --MatchOptions.gpu_index 0 --no_opengl 1 --MatchOptions.max_num_matches 40000
```

 I get the error 
```
Matching block [1/3, 1/3]
MultiplyDescriptor:	an illegal memory access was encountered
MultiplyDescriptor:	an illegal memory access was encountered
MultiplyDescriptor:	an illegal memory access was encountered
MultiplyDescriptor:	an illegal memory access was encountered
MultiplyDescriptor:	an illegal memory access was encountered
MultiplyDescriptor:	an illegal memory access was encountered
```

I tried leaving the `max_num_matches` at the default values, but the error stayed the same, except for the warning
```
WARNING: Clamping features from 37110 to 32768 - consider increasing the maximum number of matches.
```
I also tried changing the max number of features detected to be 20k instead of 40k and still had the issue.

I am using a GTX1080 on a pretty powerful Linux machine running Ubuntu 16.04.1 LTS. `nvidia-smi` output is shown below.

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 367.57                 Driver Version: 367.57                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 1080    Off  | 0000:02:00.0     Off |                  N/A |
| 27%   38C    P8    10W / 180W |      9MiB /  8113MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX 1080    Off  | 0000:03:00.0     Off |                  N/A |
| 27%   38C    P8     6W / 180W |      1MiB /  8113MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  GeForce GTX 1080    Off  | 0000:83:00.0     Off |                  N/A |
| 27%   37C    P8     6W / 180W |      1MiB /  8113MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  GeForce GTX 1080    Off  | 0000:84:00.0     Off |                  N/A |
| 27%   38C    P8     6W / 180W |      1MiB /  8113MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
```

I also tried running the command without specifying a GPU index which (from what I gather from `base/feature_matching.cc:1693:1702`) does not run CUDA and still got the error. In all cases I deleted the database between retries to make sure I was recomputing everything.

Are you sure the the CUDA file was causing the bug? It's worth noting that when I run the same exact data with the command defaults for number of features, image size, and image matching, it works just fine (although it resizes my images to 3200x1800 and only detects ~8-9k features per image).

On a different machine (running Ubuntu 14.04.5 LTS with a single GTX680) I ran the exact same data and settings and did not get this error. That Colmap build however is from an older commit

```
commit 5220eb0b7dbe6e6e48a748f5b9d9c35c6e5d6096
Author: Johannes Scho<88>nberger <jsch@demuc.de>
Date:   Thu Jan 19 13:59:36 2017 +0100
```
and I ran it using the GUI instead of from the command line.

**tl;dr the bug from #68 still seems to be a problem. However, the GUI from an older commit doesn't seem to experience it.**

**Comment by meder411** [comment]
I tried testing the same commands but with

**Comment by ahojnnes** [comment]
You are running out of GPU memory. Set --MatchOptions.max_num_matches to a lower value.

**Comment by meder411** [comment]
I'm confused how I'm running out of GPU memory. If each SIFT descriptor is 128 bytes * sizeof(float), we're looking at 1KB per descriptor. For 30k descriptors, an image should carry a 30MB descriptor set. If my block size is the default 50, I'm only looking at 1.5GB of memory used. Hell, even if I loaded all 126 images into GPU memory, we are looking at ~3.78GB of memory. My GPU has 8GB of memory. Is there a lot of other data loaded onto the GPU at the same time? I'm confused at what I'm overlooking.

**Comment by ahojnnes** [comment]
You are computing exhaustive feature matches, which means you have to store a 32768^2 float matrix.

**Comment by sbarkby** [comment]
I'm now experiencing a similar issue.

Up until now I have been using a previous commit (0a25983257f5abfa693739d1a06cd8b033fac4fc) from the 16th January. In this commit I use the default MatchOptions.max_num_matches of 32768 and it works fine. 

Now that I have updated to the most current commit I get the following error message when running the exhaustive feature matcher:

MultiplyDescriptor:	out of memory
MultiplyDescriptor:	an illegal memory access was encountered
<repeated several times>
MultiplyDescriptor:	an illegal memory access was encountered
*** Aborted at 1491578639 (unix time) try "date -d @1491578639" if you are using GNU date ***
MultiplyDescriptor:	an illegal memory access was encountered
PC: @           0xc2a87b SiftMatchCU::GetBestMatch()
MultiplyDescriptor:	an illegal memory access was encountered
MultiplyDescriptor:	an illegal memory access was encountered
MultiplyDescriptor:	an illegal memory access was encountered
MultiplyDescriptor:	an illegal memory access was encountered
*** SIGSEGV (@0x7f5f37af95a0) received by PID 10285 (TID 0x7f5e367fc700) from PID 934253984; stack trace: ***
MultiplyDescriptor:	an illegal memory access was encountered
    @     0x7f5e87d7b390 (unknown)
MultiplyDescriptor:	an illegal memory access was encountered
    @           0xc2a87b SiftMatchCU::GetBestMatch()
    @           0xc230d6 SiftMatchGPU::GetGuidedSiftMatch()
    @           0x6eabd6 colmap::MatchGuidedSiftFeaturesGPU()
    @           0x6f2a77 colmap::GuidedSiftGPUFeatureMatcher::Run()
MultiplyDescriptor:	an illegal memory access was encountered
    @           0x78393c colmap::Thread::RunFunc()
MultiplyDescriptor:	an illegal memory access was encountered
    @     0x7f5e85224c80 (unknown)
    @     0x7f5e87d716ba start_thread
MultiplyDescriptor:	an illegal memory access was encountered
    @     0x7f5e8498a82d clone
MultiplyDescriptor:	an illegal memory access was encountered
    @                0x0 (unknown)
Segmentation fault (core dumped)

I have tried several smaller MatchOptions.max_num_matches but this does not seem to solve the problem.  

Has the exhaustive matcher become more memory hungry since the 16th Jan? Or is this a bug?

**Comment by ahojnnes** [comment]
Documented here now: https://colmap.github.io/faq.html#feature-matching-fails-due-to-illegal-memory-access
